L1Trigger/TrackerTFP/src/MiniHoughTransform.cc

```diff
@@ -0,0 +1,285 @@
+#include "L1Trigger/TrackerTFP/interface/MiniHoughTransform.h"
+
+#include <numeric>
+#include <algorithm>
+#include <iterator>
+#include <deque>
+#include <vector>
+#include <set>
+#include <utility>
+#include <cmath>
+
+using namespace std;
+using namespace edm;
+using namespace trackerDTC;
+
+namespace trackerTFP {
+
+  MiniHoughTransform::MiniHoughTransform(const ParameterSet& iConfig, const Setup* setup, const DataFormats* dataFormats, int region) :
+    enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+    setup_(setup),
+    dataFormats_(dataFormats),
+    inv2R_(dataFormats_->format(Variable::inv2R, Process::ht)),
+    phiT_(dataFormats_->format(Variable::phiT, Process::ht)),
+    region_(region),
+    numBinsInv2R_(setup_->htNumBinsInv2R()),
+    numCells_(setup_->mhtNumCells()),
+    numNodes_(setup_->mhtNumDLBNodes()),
+    numChannel_(setup_->mhtNumDLBChannel()),
+    input_(numBinsInv2R_) {}
+
+  void MiniHoughTransform::consume(const TTDTC::Streams& streams) {
+    auto valid = [](int& sum, const TTDTC::Frame& frame){ return sum += (frame.first.isNonnull() ? 1 : 0); };
+    int nStubsHT(0);
+    for (int binInv2R = 0; binInv2R < numBinsInv2R_; binInv2R++) {
+      const TTDTC::Stream& stream = streams[region_ * numBinsInv2R_ + binInv2R];
+      nStubsHT += accumulate(stream.begin(), stream.end(), 0, valid);
+    }
+    stubsHT_.reserve(nStubsHT);
+    stubsMHT_.reserve(nStubsHT * numCells_);
+    for (int binInv2R = 0; binInv2R < numBinsInv2R_; binInv2R++) {
+      const int inv2R = inv2R_.toSigned(binInv2R);
+      const TTDTC::Stream& stream = streams[region_ * numBinsInv2R_ + binInv2R];
+      vector<StubHT*>& stubs = input_[binInv2R];
+      stubs.reserve(stream.size());
+      for (const TTDTC::Frame& frame : stream) {
+        StubHT* stub = nullptr;
+        if (frame.first.isNonnull()) {
+          stubsHT_.emplace_back(frame, dataFormats_, inv2R);
+          stub = &stubsHT_.back();
+        }
+        stubs.push_back(stub);
+      }
+    }
+  }
+
+  void MiniHoughTransform::produce(TTDTC::Streams& accepted, TTDTC::Streams& lost) {
+    // fill MHT cells
+    vector<deque<StubMHT*>> stubsCells(numBinsInv2R_ * numCells_);
+    for (int channel = 0; channel < numBinsInv2R_; channel++)
+      fill(input_[channel], stubsCells, channel);
+    // perform static load balancing
+    vector<vector<StubMHT*>> streamsSLB(numBinsInv2R_);
+    for (int channel = 0; channel < numBinsInv2R_; channel++) {
+      vector<deque<StubMHT*>> tmp(numCells_);
+      for (int k = 0; k < numCells_; k++)
+        swap(tmp[k], stubsCells[(channel / numCells_) * numBinsInv2R_ + channel % numCells_ + k * numCells_]);
```

Comment:
I believe this swap determines which inputs SLB muxes together. Please add comment saying so, and indicating briefly what physically it is doing. It's hard to figure out from code, unless one first understands that original ordering would correspond to channel*numCells_+k.

Reply:
done